### PR TITLE
Add SinglyLinkedList

### PR DESCRIPTION
**Pull request by daurnimator**
For my work on an allocator I needed a singly linked list.
This PR adds an implementation with #2384 in mind.

I went through current users of `std.LinkedList` in the standard library, and noticed that the `ArenaAllocator` actually wanted a singly linked list, so I changed that over as part of this PR.

### CODE REVIEW

File: std/debug/failing_allocator.zig [hunk 1]
```diff
@@ -10,6 +10,7 @@ pub const FailingAllocator = struct {
     internal_allocator: *mem.Allocator,
     allocated_bytes: usize,
     freed_bytes: usize,
+    allocations: usize,
```
Comment (andrewrk) [inline std/debug/failing_allocator.zig]: For changes to this struct, please see my review comments on #2310

File: std/heap.zig [hunk 2]
```diff
@@ -196,10 +196,10 @@ pub const ArenaAllocator = struct {
     pub allocator: Allocator,
 
     child_allocator: *Allocator,
-    buffer_list: std.LinkedList([]u8),
+    buffer_list: std.linked_list.SinglyLinkedList([]u8),
```
Comment (andrewrk) [inline std/heap.zig]: You can make it `std.SinglyLinkedList`

File: std/std.zig [hunk 3]
```diff
@@ -7,7 +7,7 @@ pub const Buffer = @import("buffer.zig").Buffer;
 pub const BufferOutStream = @import("io.zig").BufferOutStream;
 pub const DynLib = @import("dynamic_library.zig").DynLib;
 pub const HashMap = @import("hash_map.zig").HashMap;
-pub const LinkedList = @import("linked_list.zig").LinkedList;
+pub const LinkedList = @import("linked_list.zig").TailQueue;
```
Comment (andrewrk) [inline std/std.zig]: These names should match, one way or the other. I'd prefer `LinkedList` to `TailQueue`. `LinkedList` and `ArrayList` go well together.

File: std/std.zig [hunk 4]
```diff
@@ -30,6 +30,7 @@ pub const hash = @import("hash.zig");
 pub const hash_map = @import("hash_map.zig");
 pub const heap = @import("heap.zig");
 pub const io = @import("io.zig");
+pub const linked_list = @import("linked_list.zig");
```
Comment (andrewrk) [inline std/std.zig]: I liked it the other way, `std.LinkedList`.

File: std/zig/parse.zig [hunk 5]
```diff
@@ -10,13 +10,26 @@ const Error = ast.Error;
 /// Result should be freed with tree.deinit() when there are
 /// no more references to any of the tokens or nodes.
 pub fn parse(allocator: *mem.Allocator, source: []const u8) !*ast.Tree {
-    var tree_arena = std.heap.ArenaAllocator.init(allocator);
-    errdefer tree_arena.deinit();
+    const tree = blk: {
```
Comment (andrewrk) [inline std/zig/parse.zig]: What's with these seemingly unrelated changes?

These will probably conflict with the stage2 parser rewrite, I'd rather merge that before this one.

File: std/std.zig (hunk 3 above)
Comment (daurnimator) [reply std/std.zig]: The issue ( #2384 ) is that there are many types of LinkedList, and people should make an informed choice about which type they want.
I kept `std.LinkedList` as the old type to avoid making this a **breaking** PR.

File: std/std.zig (hunk 4 above)
Comment (daurnimator) [reply std/std.zig]: This allows people to write: `std.linked_list.TailQueue` vs `std.linked_list.SinglyLinkedList`, etc.
My thought was that `std.LinkedList` may get removed.

File: std/zig/parse.zig (hunk 5 above)
Comment (daurnimator) [reply std/zig/parse.zig]: See the individual commits: this fixes a bug in the parser where it incorrectly uses the allocator (it makes a copy; and then does some operations on the copy; some on the original: this gets internal allocator state confused).
@hryx knows about this (we talked on IRC), and I think he's okay with it getting merged here.

File: std/debug/failing_allocator.zig (hunk 1 above)
Comment (daurnimator) [reply std/debug/failing_allocator.zig]: > So it would introduce a new field, allocations, which increments in realloc when old_mem.len == 0. 
> Meanwhile, deallocations would increment in both realloc and shrink when new_size == 0.

This seems odd to me: why would `allocations` not increment on grow when `deallocations` does increment on shrink?

File: std/zig/parse.zig (hunk 5 above)
Comment (hryx) [reply std/zig/parse.zig]: @daurnimator It will certain conflict, so I'd rather it wait till the parser is merged. Sorry if I wasn't clear on IRC, I only meant to acknowledge the commit you linked.

File: std/heap.zig (hunk 2 above)
Comment (daurnimator) [reply std/heap.zig]: Why wouldn't we want to namespace all the various linked list alternatives under `std.linked_list`?

**Comment by daurnimator** [comment]
Discussion from IRC implemented. Should be good to merge now.
